Determine who is primary author and save e-mail in YAML field

kopeckyf opened the pull request:
Store each author's OJS ID and determine the OJS ID of the primary contact. Then check who is primary contact and output this e-mail address to YAML. 

This change is needed for the updated PDF layout which only gives the e-mail address of the primary contact.

kopeckyf committed:
Determine who is primary author and save e-mail in YAML field

thomjur commented:
Hi! Können wir gerne mergen, ich gehe davon aus, dass du das getestet hast mit verschiedenen OJS XML?

Nur zwei Fragen:
1. Können wir davon ausgehen, dass die Attribute "primary_contact_id" und "id" *immer* im XML vorhanden sind? Sieht für mich so aus, aber wenn nicht, knallt es da, weil das laut etree docs einfach ein Python Dictionary ist. Sonst vielleicht mit get() arbeiten.
2. Du könntest auch überlegen, schon ein Default-Feld im data_dict in functions.py in der entsprechenden Funktion zu erstellen.

LG
Thomas

kopeckyf commented:
`primary_contact_id` ist nicht gesetzt, wenn eine XML keine Daten zu Autoren enthält. Aber dann gibt es sowieso einen Fehler wegen Zeile 165:

https://github.com/phimisci/xml2yaml-os/blob/e0ef12f6165612a7198c9c144f9802447a974c7c/xml2yaml.py#L165

Ob ein Spezialfall ohne Autor je eintreten könnte, bespreche ich intern mit Hg.

kopeckyf committed:
Improve detection of primary contact

 - Do not rely on primary contact set in OJS
 - Store contact name, too

thomjur commented:
Ja, wahrscheinlich tritt der Fall auch nicht ein, war eher eine Rückfrage. Ich würde das etwas kürzen und vor allem die beiden Felder in die Funktion in functions.py schonmal mit default Werten (also leere Strings) aufnehmen, weil man da dann gleich eine gute Übersicht über alle Felder hat und man nicht den ganzen Code durchgehen muss, um zu schauen, welche Felder am Ende dabei sind.

Du musst das mWn nicht mit try/except abfangen, es sollte auch gehen (hab es nicht getestet:

```python
# Primary contact from OJS data
ojs_id_of_primary = publication_data.attrib.get("primary_contact_id") # Is None if field does not exist
# Loop over all authors and check who is primary contact
if ojs_id_of_primary != None:
  for a in data_dict["author"]:
      if a["ojs-id"] == ojs_id_of_primary:
          data_dict["contact_name"] = a["name"]
          data_dict["contact_email"] = a["email"]
          break
```

Wenn die Felder im data_dict bereit mit "" initialisiert wurden, würde das reichen.

thomjur commented:
Achso, das mit den asserts ist allgemein nicht so gut, muss ich mal irgendwann ändern, das stammt noch daher, dass ich den Code jedenfalls zu Anfang immer mit mypy habe statisch auf Typsicherheit habe überprüfen lassen, aber das habe ich irgendwann nicht mehr weiter verfolgt leider.

kopeckyf committed:
Update contact data storage

 - Default empty value for pandoc $if$ statement processing
 - Store as structured data
